classes/form/username_form.php

```diff
@@ -59,12 +61,18 @@ public function definition() {
     public function validation($data, $files) {
         global $CFG, $PAGE;
 
-        if (!array_key_exists('username', $data)) {
+        if (!isset($data['username']) || $data['username'] == '') {
             return array('username' => util::lang_string('form_username_not_provided'));
         }
 
         $username = $data['username'];
 
+        if (static::active_user_by_email_exists($username) && static::is_email_duplicate($username)) {
+            return array('username' => util::lang_string('duplicate_field'));
+        } elseif (static::active_user_by_email_exists($username)) {
+            return;
+        }
```

Comment:
The `static::active_user_by_email_exists($username)` is not really required here. `static::is_email_duplicate($username)` already confirms this implicitly if the return count is > 0. Also I dont think you want to return nothing in your elseif as the current functionality looks to return a bool (its a validation function).

Reply:
You may want to rework `is_email_duplicate()` to something along the lines of number_of_users_accounts() and handle the return count value to cover both of the above outcomes.